token-metadata/program/src/utils/master_edition.rs

```diff
@@ -96,49 +99,117 @@ pub fn process_mint_new_edition_from_master_edition_via_token_logic<'a>(
+    // but we can safely assume that the account was only created if those checks passed.
+    if !is_master_edition(master_edition_account_info, 0, 1) {
+        return Err(MetadataError::InvalidMasterEdition.into());
+    };
 
-    let bump = assert_derivation(
-        program_id,
-        edition_marker_info,
-        &[
-            PREFIX.as_bytes(),
-            program_id.as_ref(),
-            master_metadata.mint.as_ref(),
-            EDITION.as_bytes(),
-            as_string.as_bytes(),
-        ],
-    )?;
+    let token_standard = master_metadata
+        .token_standard
+        .unwrap_or(TokenStandard::NonFungible);
+    match token_standard {
+        TokenStandard::NonFungible => {
+            let edition_number = edition.checked_div(EDITION_MARKER_BIT_SIZE).unwrap();
+            let as_string = edition_number.to_string();
+
+            let bump = assert_derivation(
+                program_id,
+                edition_marker_info,
+                &[
+                    PREFIX.as_bytes(),
+                    program_id.as_ref(),
+                    master_metadata.mint.as_ref(),
+                    EDITION.as_bytes(),
+                    as_string.as_bytes(),
+                ],
+            )?;
+
+            if edition_marker_info.data_is_empty() {
+                let seeds = &[
+                    PREFIX.as_bytes(),
+                    program_id.as_ref(),
+                    master_metadata.mint.as_ref(),
+                    EDITION.as_bytes(),
+                    as_string.as_bytes(),
+                    &[bump],
+                ];
+
+                create_or_allocate_account_raw(
+                    *program_id,
+                    edition_marker_info,
+                    system_account_info,
+                    payer_account_info,
+                    MAX_EDITION_MARKER_SIZE,
+                    seeds,
+                )?;
+            }
 
-    if edition_marker_info.data_is_empty() {
-        let seeds = &[
-            PREFIX.as_bytes(),
-            program_id.as_ref(),
-            master_metadata.mint.as_ref(),
-            EDITION.as_bytes(),
-            as_string.as_bytes(),
-            &[bump],
-        ];
+            let mut edition_marker = EditionMarker::from_account_info(edition_marker_info)?;
+            edition_marker.key = Key::EditionMarker;
+            if edition_marker.edition_taken(edition)? {
+                return Err(MetadataError::AlreadyInitialized.into());
+            } else {
+                edition_marker.insert_edition(edition)?
+            }
+            edition_marker.serialize(&mut *edition_marker_info.data.borrow_mut())?;
+        }
+        TokenStandard::ProgrammableNonFungible => {
+            // Check that the new update authority is the same as the master edition.
+            if update_authority_info.key != &master_metadata.update_authority {
+                return Err(MetadataError::UpdateAuthorityIncorrect.into());
+            }
```

Comment:
From the PRD discussion, I was originally thinking this was a property of the new `Print` handler, not a property of pNFT specifically.  wdyt about doing it at the `Print` level?